This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [ ]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = ''

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [1]:
# First, import the relevant modules
import requests
import json

In [2]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
# Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
request = requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31api_key=')

### answer to Ques 1 above: collected data from Quandl utilizing API

In [3]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
request_text = request.text

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [4]:
# Convert the returned JSON object into a Python dictionary.
data = json.loads(request_text)

In [5]:
# Check the data type to make sure it is python object.
print(type(data))

<class 'dict'>


### answer to Ques 2 above: converted JSON object into python dictionary utilizing requests

In [6]:
# print(data)  and view in order to evaluate features of nested dictionary

In [7]:
# In order to get max and min of opening price,first view all column names
data['dataset']['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [8]:
# set columns_list equal to all the column names
columns_list = data['dataset']['column_names']

In [9]:
# set facts equal to the data stored within 'data' and corresponding to the column name indices
facts = data['dataset']['data']

In [10]:
#take a look at first two lists of lists
facts[:2]

[['2017-12-29',
  51.76,
  51.94,
  51.45,
  51.76,
  None,
  34640.0,
  1792304.0,
  None,
  None,
  None],
 ['2017-12-28',
  51.65,
  51.82,
  51.43,
  51.6,
  None,
  40660.0,
  2099024.0,
  None,
  None,
  None]]

In [11]:
# even though visible that first index is opening price, check the index of 'Open' to make sure
open_pos = columns_list.index('Open')
open_pos

1

In [12]:
# loop thru values and filter out missing values from opening prices
open_prices = [row[open_pos] for row in facts]
open_prices = list(filter(None, open_prices))

In [13]:
# now that missing values are eliminated, apply max and min function to solve answer (3):
# 3.)Calculate what the highest and lowest opening prices were for the stock in this period.
max(open_prices), min(open_prices)

(53.11, 34.0)

### Answer to Ques 3 above) highest opening price is 53.11 and lowest opening price is 34.0

In [14]:
# determine which index is associated with 'High' for highest price on any day
high_price = columns_list.index('High')
print("high_price index is: ", high_price)
# determine which index is associated with 'Low' for lowest price on any day
low_price = columns_list.index('Low')
print("low_price index is: ", low_price)
#determine which index is associated with 'Date'
date = columns_list.index('Date')
print("date index is: ", date)

high_price index is:  2
low_price index is:  3
date index is:  0


In [15]:
#loop thru all the days in list and take difference between high price and low price
#create a list of all the differences of high and low on all the dates
my_list = [d[2]-d[3] for d in facts]

In [18]:
#Ques 4 asks What was the largest change in any one day (based on High and Low price) therefore take the max of our list of differences to determine the largest change
my_list_change = max(my_list)
print(my_list_change)

2.8100000000000023


In [19]:
#determining the corresponding index number of our max difference in order to pull date from that list
corr_index = my_list.index(my_list_change)
print("corresponding index is: ", corr_index)

corresponding index is:  161


In [20]:
# in order to obtain the date that corresponds to the max difference between open and close, index the max index[161] and the date index[0]
m_diff = facts[corr_index][date]
print("date of max diff is: ",m_diff)

date of max diff is:  2017-05-11


### Answer to Ques 4 above) 2.81 is the largest change in any one day and is seen on 05/11/2017

In [21]:
# use for loop again to find largest change between any two days;first determine index of closing price.
close_pos = columns_list.index('Close')
close_pos

4

In [22]:
# loop thru values and filter out missing values
close_prices = [row[close_pos] for row in facts]
close_prices = list(filter(None, close_prices))

In [25]:
#create a loop to find greatest difference in closin prices between two consecutive days

close_prices_diff = [abs(p_next - p_prev) for (p_prev, p_next) in zip(close_prices, close_prices[1:])]
max(close_prices_diff)
print("The largest change between two consecutive days is: ", max(close_prices_diff))

The largest change between two consecutive days is:  2.559999999999995


### answer to Ques 5) the largest closing price change between two consecutive days is 2.56.

In [27]:
# to solve for ave(question 6), use sum/len and once again find index of trading volume
traded_pos = columns_list.index('Traded Volume')
traded_pos

6

In [28]:
#loop thru and first eliminate missing values
traded_vol = [row[traded_pos] for row in facts]
traded_vol = list(filter(None, traded_vol))

In [29]:
# (question 6) What was the average daily trading volume during this year?
(sum(traded_vol)) / (len(traded_vol))

89124.33725490196

### Answer to Question 6) ave daily trading volume during year is 89124.34.

In [30]:
len(traded_vol)

255

In [31]:
# use equation to determine which index is in the middle
size = len(traded_vol)
if size %2 == 1:
    median = (size / 2) + (0.5)
    
if size %2 == 0:
    median = ((size/2), (size/2 + 1))

index = median - 1
print(index)

127.0


In [32]:
#sort data 
traded_volume = sorted(traded_vol)

In [33]:
# solve question 7 by indexing the median sorted item in list
# 7)What was the median trading volume during this year. 
traded_volume[int(index)]


76286.0

### Answer to Ques 7) Median trading volume during this year is 76286.0.